[allocator] Introduce the abstract class CUDACachingAllocator

zdevito opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #87251
* #87250
* #87237

This replaces the manual function pointers, making it easier to write
new drop-in allocators.

Note that most allocation goes through the Allocator interface, which
CUDAAllocator inherits from, and this arrangement avoids adding and
additional layer of dispatch along this pathway compared to what existed before.

zdevito committed:
[allocator] Introduce the abstract class CUDACachingAllocator

This replaces the manual function pointers, making it easier to write
new drop-in allocators.

Note that most allocation goes through the Allocator interface, which
CUDAAllocator inherits from, and this arrangement avoids adding and
additional layer of dispatch along this pathway compared to what existed before.

[ghstack-poisoned]

zdevito committed:
Update on "[allocator] Introduce the abstract class CUDACachingAllocator"


This replaces the manual function pointers, making it easier to write
new drop-in allocators.

Note that most allocation goes through the Allocator interface, which
CUDAAllocator inherits from, and this arrangement avoids adding and
additional layer of dispatch along this pathway compared to what existed before.

[ghstack-poisoned]

zdevito committed:
Update on "[allocator] Introduce the abstract class CUDACachingAllocator"


This replaces the manual function pointers, making it easier to write
new drop-in allocators.

Note that most allocation goes through the Allocator interface, which
CUDAAllocator inherits from, and this arrangement avoids adding and
additional layer of dispatch along this pathway compared to what existed before.

[ghstack-poisoned]

wconstab reviewed (approved):
Mostly code-motion, i couldn't spot any typos or issues there.

Couple of nits about using_ flags  renaming

perf-wise, I agree you're swapping an equivalent load/deref from a function-pointer to a get()->.

zdevito committed:
Update on "[allocator] Introduce the abstract class CUDACachingAllocator"


This replaces the manual function pointers, making it easier to write
new drop-in allocators.

Note that most allocation goes through the Allocator interface, which
CUDAAllocator inherits from, and this arrangement avoids adding and
additional layer of dispatch along this pathway compared to what existed before.

[ghstack-poisoned]

zdevito commented:
@pytorchbot merge

zdevito committed:
Update on "[allocator] Introduce the abstract class CUDACachingAllocator"


This replaces the manual function pointers, making it easier to write
new drop-in allocators.

Note that most allocation goes through the Allocator interface, which
CUDAAllocator inherits from, and this arrangement avoids adding and
additional layer of dispatch along this pathway compared to what existed before.

[ghstack-poisoned]

zdevito commented:
@pytorchbot merge

zdevito committed:
Update on "[allocator] Introduce the abstract class CUDACachingAllocator"


This replaces the manual function pointers, making it easier to write
new drop-in allocators.

Note that most allocation goes through the Allocator interface, which
CUDAAllocator inherits from, and this arrangement avoids adding and
additional layer of dispatch along this pathway compared to what existed before.

[ghstack-poisoned]

zdevito committed:
Update on "[allocator] Introduce the abstract class CUDACachingAllocator"


This replaces the manual function pointers, making it easier to write
new drop-in allocators.

Note that most allocation goes through the Allocator interface, which
CUDAAllocator inherits from, and this arrangement avoids adding and
additional layer of dispatch along this pathway compared to what existed before.

[ghstack-poisoned]

zdevito commented:
@pytorchbot merge
